File: leetcode-js/hot100/leetcode543. Diameter of Binary Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
let map = new Map();
var diameterOfBinaryTree = function(root) {
    if(!root)
        return 0;
    if(!map.has(root.left)){
        let left = depth(root.left);
        map.set(root.left, left);
    }
    if(!map.has(root.right)){
        let right = depth(root.right);
        map.set(root.right, right);
    }
    return Math.max(diameterOfBinaryTree(root.left), diameterOfBinaryTree(root.right), map.get(root.left) + map.get(root.right));
};


function depth(root){
    if(!root)
	return 0;
    return Math.max(1 + depth(root.left), 1 + depth(root.right));
}

File: leetcode-java/tree/leetcode543. Diameter of Binary Tree.java
package tree;

import java.util.HashMap;

//Definition for a binary tree node.
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}

class Solution {
    HashMap<TreeNode, Integer> map = new HashMap<>();
    public int diameterOfBinaryTree(TreeNode root) {
        if(root == null) {
            return 0;
        }
        if(!map.containsKey(root.left)) {
            int left = depth(root.left);
            map.put(root.left, left);
        }
        if(!map.containsKey(root.right)) {
            int right = depth(root.right);
            map.put(root.right, right);
        }
        int leftNode = diameterOfBinaryTree(root.left);
        int rightNode = diameterOfBinaryTree(root.right);
        return Math.max(Math.max(leftNode, rightNode), map.get(root.left) + map.get(root.right));
    }

    public int depth(TreeNode root) {
        if(root == null) {
            return 0;
        }
        return Math.max(depth(root.left), depth(root.right)) + 1;
    }
}